### PR TITLE
[persist] Allow sorting by structured data in compaction

### PR DESCRIPTION
**Pull request by bkirwi**
- Abstract out a big chunk of the streaming consolidation code. (Mostly importantly, the KV type used for ordering is now an associated type.)
- Add the `ArrayOrd` / `ArrayIdx` abstraction which allows comparing values across arrays.
- Implement our new traits for structured data.
- Allow configuring writing structured data in the batch builder.

### Motivation

A big chunk of #24830.

### Tips for reviewer

There are a couple things we'd want to do before enabling the new compaction order widely:
- Implement a `key_lower` equivalent for structured data.
- Better memory bounding in compaction. (In particular, very long runs need to be broken up before being fetched / compacted.)
- Allow reading structured data in the read path, including fast-path peeks.
- Integrate Dan's recent data-type migration code to ensure structured data is type-correct.
- Metrics.

Punting since I think this PR is already plenty large. (And let me know if you want me to break it up further.) The current state is at least enough to enable flipping the flag around in parallel workload tests, though, so we have coverage of compacting codec-ordered batches and producing structured output as well as vice versa.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/persist-client/src/batch.rs [hunk 1]
```diff
@@ -811,51 +825,58 @@ where
...
+                    // the smallest data in the part is not regressing... but for now, keep splitting
+                    // runs as before.
+                    if (min_part_k, min_part_v, &min_part_t) < (max_run_k, max_run_v, max_run_t) {
+                        soft_panic_or_log!("expected data in sorted order");
+                        self.runs.push(self.parts_written);
+                        self.run_meta.push(
+                            self.parts
+                                .cfg
+                                .run_meta(RunOrder::Unordered, stats_schemas.id),
+                        )
```
Comment (def-) [inline src/persist-client/src/batch.rs:851]: Uncovered

**Review by def- — COMMENTED** [review]
Based on coverage report: https://buildkite.com/materialize/coverage/builds/442
Maybe we should enable structured order in all of CI? I'll do a test run with that: https://buildkite.com/materialize/test/builds/89378 & https://buildkite.com/materialize/nightly/builds/9305 Would it make sense to enable in general to get more testing coverage?

**Comment by bkirwi** [comment]
Yep, it's intentional that this is disabled in general - see "notes for reviewer" above. (There's a few more things that need to go in before this is safe to enable widely, but the parallel workload ought to work and gives us basic coverage for now.)

I promise to get through that list and enable in CI before rolling out widely in staging. Let me know if you're not comfortable with that strategy!

File: src/persist/src/indexed/encoding.rs [hunk 2]
```diff
@@ -211,6 +212,28 @@ impl BlobTraceUpdates {
             BlobTraceUpdates::Both(_, structured) => Some(structured),
         }
     }
+
+    /// Return the [`ColumnarRecordsStructuredExt`] of the blob.
+    pub fn get_or_make_structured<K: Codec, V: Codec>(
+        &mut self,
+        key_schema: &K::Schema,
+        value_schema: &V::Schema,
```
Comment (danhhz) [inline src/persist/src/indexed/encoding.rs:220]: super nit: `val_schema` 🙈

File: src/persist/src/indexed/encoding.rs [hunk 3]
```diff
@@ -211,6 +212,28 @@ impl BlobTraceUpdates {
...
+        match self {
+            BlobTraceUpdates::Row(records) => {
+                let key = codec_to_schema2::<K>(key_schema, records.keys()).expect("valid keys");
+                let val =
+                    codec_to_schema2::<V>(value_schema, records.vals()).expect("valid values");
+                *self = BlobTraceUpdates::Both(
+                    records.clone(),
+                    ColumnarRecordsStructuredExt { key, val },
+                );
+                // Recurse at most once, since this data is now structured.
```
Comment (danhhz) [inline src/persist/src/indexed/encoding.rs:231]: I have an irrational fondness for this technique

**Review by danhhz — APPROVED** [review]
I wish I had something useful to say, but going commit-by-commit it's surprisingly straightforward!

The one thing I couldn't really wrap my head around (probably because I've 100% flushed out all context) was the expected_consolidated/expected_order commit. If you're confidant that it's correct, then that's good enough for me, and feel free to merge.

LGTM

File: src/persist/src/indexed/encoding.rs (hunk 3 above)
Comment (bkirwi) [reply src/persist/src/indexed/encoding.rs:231]: 😅

File: src/repr/proptest-regressions/row/encoding2.txt [hunk 4]
```diff
@@ -11,3 +11,9 @@ cc 99bc1fecbd78903730664f137ed50776731115043d1aef6c8e7401c2c172666a # shrinks to
...
 cc 5294d4b8646bf778aadf1dd930bd9109347cc6bf51aa72105d2c85744b842586 # shrinks to (ty, datums) = (ColumnType { scalar_type: Jsonb, nullable: false }, [])
 cc d66b43db292454632fa92406fb76800ddb9e1ce4ab11e86e6f3d886d1db6c52a # shrinks to (ty, datums) = (ColumnType { scalar_type: MzAclItem, nullable: false }, [])
+cc f065b348d11aee106ee8e1181d00a2eea34f0b2ee5ac346223eed55439fcef50 # shrinks to (ty, datums) = (ColumnType { scalar_type: Interval, nullable: false }, [Interval(Interval { months: 0, days: 0, micros: 0 }), Interval(Interval { months: 0, days: 0, micros: 0 })])
```
[the remark this answers is not in the document]
Comment (bkirwi) [reply src/repr/proptest-regressions/row/encoding2.txt:14]: I believe so yeah! (eg. This line is me not adding fixed-size array support yet.)

**Comment by bkirwi** [comment]
> The one thing I couldn't really wrap my head around (probably because I've 100% flushed out all context) was the expected_consolidated/expected_order commit.

Cool - I'll type up the thinking here, and we can follow up next week if anything doesn't scan.

Talking about [this commit](https://github.com/MaterializeInc/materialize/pull/29121/commits/7275a3cbe66bb0f6f914d2c26892d4638646494c):
- Historically the batch builder expects to get data in two orders: unordered (for user batches) and codec-ordered (for compaction). We're adding a third possible order, so we want to swap the boolean for our new enum.
- It's very important that compaction and the batch builder agree on the order, even if the flag changes dynamically. So we write down the order in the batch-builder config, and condition on that everywhere. This means that the expect_ordered boolean now moves to the batch config constructor instead of the batch builder.
- The worst bug I had when developing this was: historically the batch builder decides the key_lower is the first entry in the keys array when the data is ordered. This is no longer true in structured-data order, and choosing the wrong key lower causes data to be read back in the wrong order. Pushing the enum into the batch builder helped force thinking about all three cases...

**Pull request merged**